sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -144,6 +144,15 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
           value != null =>
       Some(unwrapTimestampToDate(be, fromExp, ts, timeZoneId, evalMode))
 
+    // Timestamp/Timestamp_NTZ -> Timestamp_NTZ/Timestamp
+    case be @ BinaryComparison(
+      Cast(fromExp, _, timeZoneId, evalMode), Literal(value, literalType))
+        if AnyTimestampType.acceptsType(fromExp.dataType) &&
+          AnyTimestampType.acceptsType(literalType) && value != null =>
+      val newExpr = be.withNewChildren(Seq(fromExp, Cast(Literal(value, literalType),
+        fromExp.dataType, timeZoneId, evalMode)))
+      Some(newExpr)
```

Comment:
Please let me know if this makes sense (if I miss anything this could be wrong).

Reply:
cc @sunchao

Reply:
LGTM.